sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/SpringSdkMessageCodec.scala

```diff
@@ -30,10 +32,59 @@ private[springsdk] object SpringSdkMessageCodec extends MessageCodec {
    * In the Spring SDK, output data are encoded to Json.
    */
   override def encodeScala(value: Any): ScalaPbAny =
-    ScalaPbAny.fromJavaProto(encodeJava(value))
+    ScalaPbAny.fromJavaProto(JsonSupport.encodeJson(value, findTypeHint(value)))
 
   override def encodeJava(value: Any): JavaPbAny =
-    JsonSupport.encodeJson(value)
+    JsonSupport.encodeJson(value, findTypeHint(value))
+
+  private def findTypeHint(value: Any): String = {
+
+    // TODO: cache results, otherwise we'll run scanning + reflection on each encoding
+    def annotatedParents(clz: Class[_], listOfParents: Seq[Class[_]]): Seq[Class[_]] = {
+
+      def hasJsonSubTypes(clz: Class[_]) =
+        clz != null && clz.getAnnotation(classOf[JsonSubTypes]) != null
+
+      val acc =
+        if (hasJsonSubTypes(clz)) listOfParents :+ clz
+        else listOfParents
+
+      val directParents = clz.getSuperclass +: clz.getInterfaces
+
+      directParents.foldLeft(acc) { case (acc, clz) =>
+        if (clz == null) acc // happens when we reach the bottom, ie: Object.getSuperclass == null
+        else annotatedParents(clz, acc)
+      }
+    }
+
+    // straightforward case: class is annotated with JsonTypeName
```

Comment:
I really like these comments. They show the intent which is not often obvious. I think we would be better off with more of them. Would it be worthy to discuss this as a team?

Reply:
I don't think that's worth. It's more a matter of style and taste.